// @adobe/pdfservices-node-sdk/src/internal/internal-execution-context.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const InternalClientConfig = require('./internal-client-config'),
	AuthenticatorFactory = require('./auth/authenticator-factory');
	ServiceAccountCredentials = require('../auth/service-account-credentials'),
	ServiceTokenCredentials = require('./auth/service-token-credentials'),
	HttpRequest = require('../internal/http/http-request'),
	CPFServiceRequestContext = require('./cpf/cpf-service-request-context'),
	ServiceAccountCredentialsWithUri = require('../internal/auth/service-account-credentials-with-uri');


class InternalExecutionContext{
	constructor(credentials, clientConfig){

		if (credentials instanceof ServiceAccountCredentials  || credentials instanceof ServiceTokenCredentials) {
			const opsCreateUri = credentials.getOpsCreateUri();
			if (clientConfig instanceof InternalClientConfig)
				this.clientConfig = clientConfig;
			else
				this.clientConfig = new InternalClientConfig();
			this.clientConfig.validate();
			this.credentials = credentials;
			this.credentials.validate();
			this.authenticator = AuthenticatorFactory.getAuthenticator(credentials);
			this.cpfServiceRequestContext = new CPFServiceRequestContext(
				(opsCreateUri != null)?
			this.credentials : this.clientConfig);
		}
		else {
			throw new Error("Invalid Credentials provided as argument");
		}
	}
	getClientConfig() {
		return this.clientConfig;
	}

	getBaseRequestFromRequestContext(requestKey) {
		let baseRequest = this.cpfServiceRequestContext.getBaseHttpRequest(requestKey);
		baseRequest = baseRequest.withAuthenticator(this.authenticator);
		return Promise.resolve(baseRequest);
	}


	validate(){
		this.clientConfig.validate();
		if (this.authenticator == null) {
			throw new Error("Authentication not initialized in the provided context");
		}
	}

}

module.exports = InternalExecutionContext;
