// @adobe/pdfservices-node-sdk/src/internal/config/cpf-engine-config.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const cpfConfig = {
	createPdf: {
		operation: 'Create PDF',
		assetId: 'urn:aaid:cpf:Service-1538ece812254acaac2a07799503a430'
	},
	combinePdf: {
		operation: 'Combine PDF',
		assetId: 'urn:aaid:cpf:Service-916ee91c156b42349a7847a7d564fb13'
	},
	exportPdf: {
		operation: 'Export PDF',
		assetId: 'urn:aaid:cpf:Service-26c7fda2890b44ad9a82714682e35888'
	},
	htmlToPdf: {
		operation: 'HTML to PDF',
		assetId: 'urn:aaid:cpf:Service-e2ee120a2b06427cb449592f5db967e7'
	},
	ocr: {
		operation: 'OCR',
		assetId: 'urn:aaid:cpf:Service-7e6a5d2b6bb141d7832398076914a07b'
	},
	compressPdf: {
		operation: 'Compress PDF',
		assetId: 'urn:aaid:cpf:Service-f37d36f4e6724eed92149a8ff35ea061'
	},
	linearizePdf: {
		operation: 'Linearize PDF',
		assetId: 'urn:aaid:cpf:Service-e4d5f0b75e5d43ea9eaa187860772d27'
	},
	//PAPI format
	protectPdf: {
		assetId : 'urn:aaid:cpf:Service-627aa455ac5d4c338ad49ca882e0fce5'
	},
	pageManipulation:{
		assetId : 'urn:aaid:cpf:Service-4735fcf3cf924b25879e6fcf7aa84ad4'
	},
	removeProtection:{
		assetId : 'urn:aaid:cpf:Service-d80dc37d3aee44a8839e3429360db9a7'
	},
	splitPdf: {
		assetId : 'urn:aaid:cpf:Service-d99c7660cba14e5c98f9023221dab40f'
	},
	mergeDocument: {
		assetId: 'urn:aaid:cpf:Service-52d5db6097ed436ebb96f13a4c7bf8fb'
	},
	extractPdf: {
		operation: 'Extract PDF',
		assetId : 'urn:aaid:cpf:58af6e2c-1f0c-400d-9188-078000185695'
	},
	PDFProperties: {
		assetId : 'urn:aaid:cpf:Service-fd9b06fe2f164df7975254581d6ab00e'
	},
	formatKeyString:"dc:format",
	multipartKeyString:'sensei:multipart_field_name'

};

module.exports = cpfConfig;
