// @adobe/pdfservices-node-sdk/src/operation/combine-file-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const CombinePDFService = require('../internal/api/combine-pdf-service'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	CombineOperationInput = require('./../internal/io/combine-operation-input'),
	OperationMessage = require('../internal/cpf/operation-message'),
	cpfConfig = require('../internal/config/cpf-engine-config.js'),
	PageRanges = require('./option/page-ranges'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileWithPageOptions
	} = require('./../internal/util/validation-util');


/**
 *
 * Supported source file formats for {@link CombineFilesOperation}.
 * @enum
 * @readonly
 * @memberOf CombineFilesOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		},

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension

	};

Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * Combines multiple PDF files into a single PDF file. Allows specifying which pages of the source files to combine.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        CombineFiles = PDFServicesSdk.CombineFiles,
 *        combineFilesOperation = CombineFiles.Operation.createNew(),
 *        input1 = PDFServicesSdk.FileRef.createFromLocalFile('test/resources/combinePagesInput1.pdf', CombineFiles.SupportedSourceFormat.pdf),
 *        input2 = PDFServicesSdk.FileRef.createFromLocalFile('test/resources/combinePagesInput1.pdf', CombineFiles.SupportedSourceFormat.pdf);
 *
 *  combineFilesOperation.addInput(input1);
 *  combineFilesOperation.addInput(input2);
 *
 *  combineFilesOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/CombinedPDF.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class CombineFilesOperation {
	/**
     * @hideconstructor
     */
	constructor() {
		this.filesToCombine = [];
		Object.preventExtensions(this);
	}

	/**
     *
     * Constructs a {@link CombineFilesOperation} instance.
     * @returns {CombineFilesOperation} A new CombineFilesOperation instance.
     *
     */
	static createNew() {
		return new CombineFilesOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/* eslint no-param-reassign:0 */
	/**
     * Specifies particular pages of a PDF file (media type "application/pdf") to be combined with other files. The pages
     * will be added after the pages of any previously specified files. If the <code>pageRanges</code> argument is not
     * provided, all pages of the PDF will be added in the combined PDF.
     * <p>
     *
     * @param {!FileRef} sourceFileRef - A PDF file to be combined.
     * @param {PageRanges=} pageRanges - Page ranges of the PDF file to be combined.
     */
	addInput(sourceFileRef, pageRanges) {
		if (!pageRanges) {
			pageRanges = new PageRanges();
			pageRanges.addAll();
		}
		this.filesToCombine.push(CombineOperationInput.createNew(sourceFileRef, pageRanges));
	}

	/**
     * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
     *
     * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
     * to the actual path).
     * See {@link FileRef} for how temporary resources are cleaned up.
     *
     * @param {!ExecutionContext} context - The context in which the operation will be executed.
     * @returns {Promise<FileRef>} A promise which resolves to the operation result.
     * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
     */
	execute(context){
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}

		logger.info('All validations successfully done. Beginning Combine Operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.getSourceFileRefs(), targetFileName, DefaultConfig.operationName.combinePdf),
			combinePDFService = new CombinePDFService();
		operationMessage.setIncludeRanges(this.getIncludeRanges());
		return new Promise((resolve, reject) => {
			combinePDFService.perform(context, operationMessage)
				.then(res => resolve(res))
				.catch(err => reject(err));
		});
	}

	validate(context) {
		validateClientContext(context);
		if (this.filesToCombine.length === 0) {
			throw new Error('No input was provided for combining files');
		}
		this.filesToCombine.forEach(file => {
			validateFileWithPageOptions(file, allowedConfiguration.getSupportedMediaTypes());
		});
		if (this.filesToCombine.length > 20) {
			throw new Error('Only 20 input files can be combined in one combine operation instance');
		}
	}

	getIncludeRanges() {
		const includeRanges = [];
		this.filesToCombine.forEach(file => includeRanges.push(file.getPageRanges().getRanges()));
		return includeRanges;
	}
	getSourceFileRefs(){
		let sourceFileRefs = [];
		this.filesToCombine.forEach(file => sourceFileRefs.push(file.getSourceFileRef()));
		return sourceFileRefs;
	}
}

Object.freeze(CombineFilesOperation);
module.exports = CombineFilesOperation;
