// @adobe/pdfservices-node-sdk/src/operation/compress-pdf-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const CompressPDFService = require('../internal/api/compress-pdf-service'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	OperationMessage = require('../internal/cpf/operation-message'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	CompressPDFOptions = require('./option/compresspdf/compress-pdf-options'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType,
		validateOptionInstanceType, validateOperationOptions
	} = require('./../internal/util/validation-util');

/**
 *
 * Supported source file formats for {@link CompressPDFOperation}.
 * @enum
 * @readonly
 * @memberOf CompressPDFOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension,

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		},
		optionInstanceMapping: new Map([
			[ExtensionMediaTypeMapping.pdf, CompressPDFOptions]
		])
	};

Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An operation that reduces the size of a PDF file.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        CompressPDF = PDFServicesSdk.CompressPDF,
 *        compressPdfOperation = CompressPDF.Operation.createNew(),
 *        input = PDFServicesSdk.FileRef.createFromLocalFile('test/resources/compressInput.pdf', CompressPDF.SupportedSourceFormat.pdf);
 *
 *  compressPdfOperation.setInput(input);
 *
 *  compressPdfOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/compressOutput.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class CompressPDFOperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		this.options = null;
		Object.preventExtensions(this);
	}

	/**
	 * Constructs a {@link CompressPDFOperation} instance.
	 * @returns {CompressPDFOperation} A new CompressPDFOperation instance.
	 *
	 */
	static createNew() {
		return new CompressPDFOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Sets the options for this operation. See {@link CompressPDFOptions} for how to specify the
	 * options for the different compression levels.
	 * @param {Object=} options - optional parameters.
	 */
	setOptions(options) {
		this.options = options;
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */

	execute(context) {
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}

		logger.info('All validations successfully done. Beginning Compress PDF operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.compressPdf),
			compressPDFService = new CompressPDFService();
		operationMessage.setOptions(this.options);
		return compressPDFService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);
		if (this.options != null) {
			validateOptionInstanceType(allowedConfiguration.optionInstanceMapping,this.sourceFileRef,this.options);
			validateOperationOptions(this.options);
		}
	}

}

Object.freeze(CompressPDFOperation);
module.exports = CompressPDFOperation;
