// @adobe/pdfservices-node-sdk/src/operation/delete-pages-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const DeletePagesService = require('../internal/api/delete-pages-service'),
	OperationMessage = require('../internal/cpf/operation-message'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	PageActions = require('../internal/cpf/request/platform/page-actions'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType, validatePageRangesForOperation
	} = require('./../internal/util/validation-util');


/**
 *
 * Supported source file formats for {@link DeletePagesOperation}.
 * @enum
 * @readonly
 * @memberOf DeletePagesOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension,

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		}
	};
Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An operation to delete pages in a PDF file.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        DeletePages = PDFServicesSdk.DeletePages,
 *        deletePagesOperation = DeletePages.Operation.createNew(),
 *        pageRanges = new PDFServicesSdk.PageRanges();
 *
 *  deletePagesOperation.setInput( PDFServicesSdk.FileRef.createFromLocalFile('~/Documents/deletePagesOperationInput.pdf',
 *  	DeletePages.SupportedSourceFormat.pdf));
 *  pageRanges.addSinglePage(1);
 *  deletePagesOperation.setPageRanges(pageRanges);
 *
 *  deletePagesOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/deletePagesOperationOutput.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class DeletePagesOperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		this.options = null;
		Object.preventExtensions(this);
	}

	/**
	 * Constructs a {@link DeletePagesOperation} instance.
	 * @returns {DeletePagesOperation} A new DeletePagesOperation instance.
	 *
	 */
	static createNew() {
		return new DeletePagesOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Specifies the pages to delete from the input PDF file.
	 *
	 * @param {!PageRanges} pageRanges - Page ranges for deletion.
	 */
	setPageRanges(pageRanges){
		validatePageRangesForOperation(pageRanges);
		this.options = new PageActions().withDeleteAction(pageRanges.getRanges());
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */

	execute(context) {
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}
		logger.info('All validations successfully done. Beginning Delete Pages operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.deletePages),
			deletePagesService = new DeletePagesService();
		operationMessage.setOptions(this.options);
		return deletePagesService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);
		if(!this.options)
			throw new Error('No page ranges were set for the operation');
	}

}

Object.freeze(DeletePagesOperation);
module.exports = DeletePagesOperation;
