// @adobe/pdfservices-node-sdk/src/operation/insert-pages-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const InsertPagesService = require('../internal/api/insert-pages-service'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	CombineOperationInput = require('./../internal/io/combine-operation-input'),
	OperationMessage = require('../internal/cpf/operation-message'),
	PageRanges = require('./option/page-ranges'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType, validatePageRangesForOperation
	} = require('./../internal/util/validation-util');


/**
 *
 * Supported source file formats for {@link InsertPagesOperation}.
 * @enum
 * @readonly
 * @memberOf InsertPagesOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		},

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension

	};

Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An operation that can be used to insert pages of multiple PDF files into a base PDF file.
 * <p>
 * For more complex use cases, refer the {@link CombineFilesOperation}
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        InsertPages = PDFServicesSdk.InsertPages,
 *        insertPagesOperation = InsertPages.Operation.createNew();
 *
 *  insertPagesOperation.setBaseInput(PDFServicesSdk.FileRef.createFromLocalFile('~/Documents/insertPagesOperationBaseInput.pdf',
 *  	InsertPages.SupportedSourceFormat.pdf));
 *  insertPagesOperation.addPagesToInsertAt(1, PDFServicesSdk.FileRef.createFromLocalFile('~/Documents/insertPagesOperationFileToInsertInput.pdf',
 *  	InsertPages.SupportedSourceFormat.pdf));
 *
 *  insertPagesOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/insertPagesOperationOutput.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class InsertPagesOperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.filesToInsert = {};
		this.baseSourceFileRef = null;
		Object.preventExtensions(this);
	}

	/**
	 *
	 * Constructs a {@link InsertPagesOperation} instance.
	 * @returns {InsertPagesOperation} A new InsertPagesOperation instance.
	 *
	 */
	static createNew() {
		return new InsertPagesOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets a base input file.
	 * @param {!FileRef} baseSourceFileRef - The base input file.
	 */
	setBaseInput(baseSourceFileRef){
		this.baseSourceFileRef = baseSourceFileRef;
	}

	/**
	 * Adds the pages of the input PDF file to be inserted at the specified page of the base PDF file.
	 * <p>
	 * This method can be invoked multiple times with the same or different input PDF files.
	 * <p>
	 * If <code>pageRanges</code> is not provided, all the pages of the input PDF file will be inserted at the specified
	 * page of the base PDF file.
	 *
	 * @param {!Number} basePage - Page of the base PDF file.
	 * @param {!FileRef} inputFile - A PDF file for insertion.
	 * @param {PageRanges=} pageRanges - Page ranges of the input PDF file; defaults to all the pages.
	 */
	addPagesToInsertAt(basePage, inputFile, pageRanges ){
		if (!pageRanges) {
			pageRanges = new PageRanges();
			pageRanges.addAll();
		}
		validateFileRef(inputFile);
		validatePageRangesForOperation(pageRanges);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), inputFile);

		if(basePage < 1)
			throw new Error("Base file page should be greater than 0");

		let combineOperationInput = CombineOperationInput.createNew(inputFile, pageRanges);
		this.updateFilesToInsert(basePage, combineOperationInput);
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */
	execute(context){
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}

		logger.info('All validations successfully done. Beginning Insert Pages Operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.filesToInsert, targetFileName, DefaultConfig.operationName.insertPages),
			insertPagesService = new InsertPagesService();
		operationMessage.setBaseFileRef(this.baseSourceFileRef);

		return new Promise((resolve, reject) => {
			insertPagesService.perform(context, operationMessage)
				.then(res => resolve(res))
				.catch(err => reject(err));
		});
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.baseSourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.baseSourceFileRef);
		if(Object.keys(this.filesToInsert).length === 0)
			throw new Error('No files to insert in the base input file')
	}

	updateFilesToInsert(basePage, combineOperationInput){
		if (this.filesToInsert.hasOwnProperty(basePage)) {
			let inputList = this.filesToInsert[basePage];
			inputList.push(combineOperationInput);
		}
		else {
			let inputList = [];
			inputList.push(combineOperationInput);
			this.filesToInsert[basePage] = inputList;
		}
	}
}

Object.freeze(InsertPagesOperation);
module.exports = InsertPagesOperation;
