// @adobe/pdfservices-node-sdk/src/operation/linearize-pdf-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const LinearizePDFService = require('../internal/api/linearize-pdf-service'),
	OperationMessage = require('../internal/cpf/operation-message'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType
	} = require('./../internal/util/validation-util');

/**
 *
 * Supported source file formats for {@link LinearizePDFOperation}.
 * @enum
 * @readonly
 * @memberOf LinearizePDFOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension,

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		}
	};

Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An operation that converts a PDF file into a linearized (also known as “web optimized”) PDF file.
 * Such PDF files are optimized for incremental access in network environments.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        LinearizePDF = PDFServicesSdk.LinearizePDF,
 *        linearizePdfOperation = LinearizePDF.Operation.createNew(),
 *        input = PDFServicesSdk.FileRef.createFromLocalFile('test/resources/linearizeInput.pdf', LinearizePDF.SupportedSourceFormat.pdf);
 *
 *  linearizePdfOperation.setInput(input);
 *
 *  linearizePdfOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/linearizeOutput.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class LinearizePDFOperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		Object.preventExtensions(this);
	}

	/**
	 * Constructs a {@link LinearizePDFOperation} instance.
	 * @returns {LinearizePDFOperation} A new LinearizePDFOperation instance.
	 *
	 */
	static createNew() {
		return new LinearizePDFOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */

	execute(context) {
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}

		logger.info('All validations successfully done. Beginning Linearize PDF operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.linearizePdf),
			linearizePDFService = new LinearizePDFService();

		return linearizePDFService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);
	}

}

Object.freeze(LinearizePDFOperation);
module.exports = LinearizePDFOperation;
