// @adobe/pdfservices-node-sdk/src/operation/ocr-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const OcrService = require('../internal/api/ocr-service'),
	OperationMessage = require('../internal/cpf/operation-message'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	OcrOptions = require('./option/ocr/ocr-options'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType,
		validateOptionInstanceType, validateOperationOptions
	} = require('./../internal/util/validation-util');


/**
 *
 * Supported source file formats for {@link OCROperation}.
 * @enum
 * @readonly
 * @memberOf OCROperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension,

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		},
		optionInstanceMapping: new Map([
										   [ExtensionMediaTypeMapping.pdf, OcrOptions]
									   ])

	};
Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * Convert a PDF file into a searchable PDF file. Allows specifying locale({@link OCROptions.OCRSupportedLocale}) and
 * OCR Type ({@link OCROptions.OCRSupportedType}) for performing OCR (Optical Character Recognition)
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        OCR = PDFServicesSdk.OCR,
 *        ocrOperation = OCR.Operation.createNew(),
 *        input = PDFServicesSdk.FileRef.createFromLocalFile('test/resources/ocrInput.pdf', OCR.SupportedSourceFormat.pdf);
 *
 *  ocrOperation.setInput(input);
 *
 *  ocrOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/ocrOutput.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class OCROperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		this.options = null;
		Object.preventExtensions(this);
	}

	/**
	 * Constructs a {@link OCROperation} instance.
	 * @returns {OCROperation} A new OCROperation instance.
	 *
	 */
	static createNew() {
		return new OCROperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Sets the options for this operation. See {@link OCROptions} for how to specify the options for the different locales and
	 * ocr types.
	 * @param {Object=} options - optional parameters; use null for default values.
	 */
	setOptions(options) {
		this.options = options;
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */

	execute(context) {
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}
		logger.info('All validations successfully done. Beginning OCR operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.ocr),
			ocrService = new OcrService();
		operationMessage.setOptions(this.options);
		return ocrService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);

		if (this.options != null) {
			validateOptionInstanceType(allowedConfiguration.optionInstanceMapping,this.sourceFileRef,this.options);
			validateOperationOptions(this.options);
		}
	}

}

Object.freeze(OCROperation);
module.exports = OCROperation;
