// @adobe/pdfservices-node-sdk/src/operation/pdf-properties-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */


const PDFPropertiesService = require('../internal/api/pdf-properties-service'),
	OperationMessage = require('../internal/cpf/operation-message'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{getRandomFileNameWithExtension} = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	outputType = require('../internal/io/output-types'),
	DefaultConfig = require('../internal/config/dc-services-default-config.js'),
	PDFPropertiesOptions = require('./option/pdfproperties/pdf-properties-options.js'),
	{
		validateFileRef, validateAllowedMediaType, validateOptionInstanceType, validateOperationOptions
	} = require('./../internal/util/validation-util');

/**
 *
 * Supported source file formats for {@link PDFPropertiesOperation}.
 * @enum
 * @readonly
 * @memberOf PDFPropertiesOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},
	allowedConfiguration = {

		targetFileExtension: ExtensionMediaTypeMapping.json.extension,

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		},

		optionInstanceMapping: new Map([
			[ExtensionMediaTypeMapping.pdf, PDFPropertiesOptions]
		])
	};
Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An Operation that is used to fetch properties from an input PDF file. The properties are returned in a JSON file.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>const credentials =  PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *          .fromFile("pdfservices-api-credentials.json")
 *          .build(),
 *       clientContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *       PDFProperties = PDFServicesSdk.PDFProperties,
 *       pdfPropertiesOperation = PDFProperties.Operation.createNew(),
 *       input = PDFServicesSdk.FileRef.createFromLocalFile('test/resources/PDFPropertiesOperationInput.pdf',PDFProperties.SupportedSourceFormat.pdf);
 *
 * pdfPropertiesOperation.setInput(input);
 *
 * pdfPropertiesOperation.execute(clientContext)
 * 	.then(result => console.log(result))
 * 	.catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class PDFPropertiesOperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		this.isInvoked = false;
		this.options = null;
		Object.preventExtensions(this);

	}

	/**
	 * Constructs a {@link PDFPropertiesOperation} instance.
	 * @returns {PDFPropertiesOperation} A new PDFPropertiesOperation instance.
	 */
	static createNew() {
		return new PDFPropertiesOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Sets the options for this operation. See {@link PDFPropertiesOptions} for how to specify options for PDF Properties.
	 *
	 * @param {PDFPropertiesOptions=} options - Options specifying the parameters for this operation; use null for default values.
	 */
	setOptions(options) {
		this.options = options;
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */

	executeAndReturnFileRef(context) {
		try {
			this.validate();
		} catch (err) {
			return Promise.reject(err);
		}
		logger.info('All validations successfully done. Beginning PDF Properties operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.PDFProperties),
			pdfPropertiesService = new PDFPropertiesService();
		this.isInvoked = true;
		operationMessage.setOptions(this.options);
		return pdfPropertiesService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<Object>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */
	execute(context) {
		try {
			this.validate();
		} catch (err) {
			return Promise.reject(err);
		}
		logger.info('All validations successfully done. Beginning PDF Properties operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.PDFProperties),
			pdfPropertiesService = new PDFPropertiesService();
		this.isInvoked = true;
		operationMessage.setOutputType(outputType.jsonObject);
		operationMessage.setOptions(this.options);
		return pdfPropertiesService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	validate() {
		if (this.isInvoked)
			throw new Error('Operation instance must only be invoked once');
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);
		if (this.options != null) {
			validateOptionInstanceType(allowedConfiguration.optionInstanceMapping,this.sourceFileRef,this.options);
			validateOperationOptions(this.options);
		}

	}

}

Object.freeze(PDFPropertiesOperation);
module.exports = PDFPropertiesOperation;
