// @adobe/pdfservices-node-sdk/src/operation/reorder-pages-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const CombinePDFService = require('../internal/api/combine-pdf-service'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	OperationMessage = require('../internal/cpf/operation-message'),
	PageRanges = require('./option/page-ranges'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType, validatePageRangesForOperation
	} = require('./../internal/util/validation-util');


/**
 *
 * Supported source file formats for {@link ReorderPagesOperation}.
 * @enum
 * @readonly
 * @memberOf ReorderPagesOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		},

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension

	};

Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An operation that allows to rearrange pages in a PDF file according to the specified order.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        ReorderPages = PDFServicesSdk.ReorderPages,
 *        reorderPagesOperation = ReorderPages.Operation.createNew(),
 *        pageRanges = new PDFServicesSdk.PageRanges();
 *
 *  reorderPagesOperation.setInput(PDFServicesSdk.FileRef.createFromLocalFile('~/Documents/reorderPagesOperationInput.pdf',
 *  	ReorderPages.SupportedSourceFormat.pdf));
 *  pageRanges.addSinglePage(3);
 *  pageRanges.addPageRange(1,2);
 *  reorderPagesOperation.setPagesOrder(pageRanges);
 *
 *  reorderPagesOperation.execute(executionContext)
 *      .then(result => result.saveAsFile('output/reorderPagesOperationOutput.pdf'))
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class ReorderPagesOperation {
	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		this.pageRanges = null;
		Object.preventExtensions(this);
	}

	/**
	 *
	 * Constructs a {@link ReorderPagesOperation} instance.
	 * @returns {ReorderPagesOperation} A new ReorderPagesOperation instance.
	 *
	 */
	static createNew() {
		return new ReorderPagesOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}

	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Sets the order of the pages.
	 *
	 * @param {!PageRanges} pageRanges - Page ranges for reordering.
	 */
	setPagesOrder(pageRanges){
		this.pageRanges = pageRanges;
	}

	/**
	 * Executes this operation using the supplied context and returns a Promise which resolves to the operation result.
	 *
	 * The resulting file may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */
	execute(context){
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}

		logger.info('All validations successfully done. Beginning Reorder Pages Operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.getSourceFileRefs(), targetFileName, DefaultConfig.operationName.reorderPages),
			combinePDFService = new CombinePDFService();
		operationMessage.setIncludeRanges(this.getIncludeRanges());
		return new Promise((resolve, reject) => {
			combinePDFService.perform(context, operationMessage)
				.then(res => resolve(res))
				.catch(err => reject(err));
		});
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);
		validatePageRangesForOperation(this.pageRanges);
	}

	getIncludeRanges() {
		const includeRanges = [];
		includeRanges.push(this.pageRanges.getRanges());
		return includeRanges;
	}
	getSourceFileRefs(){
		let sourceFileRefs = [];
		sourceFileRefs.push(this.sourceFileRef);
		return sourceFileRefs;
	}
}

Object.freeze(ReorderPagesOperation);
module.exports = ReorderPagesOperation;
