// @adobe/pdfservices-node-sdk/src/operation/split-pdf-operation.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const SplitPDFService = require('../internal/api/split-pdf-service'),
	OperationMessage = require('../internal/cpf/operation-message'),
	ExtensionMediaTypeMapping = require('../internal/extension-mediatype-mapping'),
	{ getRandomFileNameWithExtension } = require('../internal/util/path-util'),
	logger = require('./../internal/logger'),
	PageRanges = require('./option/page-ranges'),
	DefaultConfig =  require('../internal/config/dc-services-default-config.js'),
	{
		validateClientContext, validateFileRef, validateAllowedMediaType, validateSplitOptions
	} = require('./../internal/util/validation-util');


/**
 *
 * Supported source file formats for {@link SplitPDFOperation}.
 * @enum
 * @readonly
 * @memberOf SplitPDFOperation
 *
 */
const SupportedSourceFormat = {

		/**
		 * Represents "application/pdf" media type
		 * @type {string}
		 */
		pdf: ExtensionMediaTypeMapping.pdf.mediaType
	},

	allowedConfiguration = {

		targetFileExtension: ExtensionMediaTypeMapping.pdf.extension,

		getSupportedMediaTypes() {
			return Object.values(SupportedSourceFormat);
		}
	};
Object.freeze(allowedConfiguration);
Object.freeze(SupportedSourceFormat);

/**
 * An operation that splits PDF document into multiple smaller documents by simply specifying either the number of files,
 * pages per file, or page ranges.
 *
 * Sample Usage:
 * <pre class="prettyprint">
 * <code>
 *  const credentials = PDFServicesSdk.Credentials.serviceAccountCredentialsBuilder()
 *            .fromFile("pdfservices-api-credentials.json")
 *            .build(),
 *        executionContext = PDFServicesSdk.ExecutionContext.create(credentials),
 *        SplitPDF = PDFServicesSdk.SplitPDF,
 *        splitPDFOperation = SplitPDF.Operation.createNew();
 *
 *  splitPDFOperation.setInput(PDFServicesSdk.FileRef.createFromLocalFile('~/Documents/splitPDFInput.pdf',
 *  	SplitPDF.SupportedSourceFormat.pdf));
 *  splitPDFOperation.setPageCount(2);
 *
 *  splitPDFOperation.execute(executionContext)
 *      .then(result => {
 *           let saveFilesPromises = [];
 *           for(let i = 0; i < result.length; i++){
 *               saveFilesPromises.push(result[i].saveAsFile(`output/splitPDFOutput_{i}.pdf`));
 *           }
 *           return Promise.all(saveFilesPromises);
 *       })
 *      .catch(err => console.log(err));
 * </code>
 * </pre>
 *
 */
class SplitPDFOperation {

	/**
	 * @hideconstructor
	 */
	constructor() {
		this.sourceFileRef = null;
		this.pageRanges = null;
		this.pageCount = null;
		this.fileCount = null;
		Object.preventExtensions(this);
	}

	/**
	 * Constructs a {@link SplitPDFOperation} instance.
	 * @returns {SplitPDFOperation} A new SplitPDFOperation instance.
	 *
	 */
	static createNew() {
		return new SplitPDFOperation();
	}

	static get SupportedSourceFormat() {
		return SupportedSourceFormat;
	}


	/**
	 * Sets an input file.
	 * @param {!FileRef} sourceFileRef - An input file.
	 */
	setInput(sourceFileRef) {
		this.sourceFileRef = sourceFileRef;
	}

	/**
	 * Sets the page ranges on the basis of which to split the input PDF file.
	 * Each page range corresponds to a single output file having the pages specified in the page range.
	 *
	 * @param {!PageRanges} pageRanges - Page ranges for split.
	 */
	setPageRanges(pageRanges){
		this.pageRanges = pageRanges;
	}

	/**
	 * Sets the number of documents to split the input PDF.
	 *
	 * @param {!Number} fileCount - Represents the number of output files.
	 */
	setFileCount(fileCount){
		this.fileCount = fileCount
	}

	/**
	 * Sets the maximum number of pages each of the output files can have.
	 *
	 * @param {!Number} pageCount - Maximum number of pages per output file.
	 */
	setPageCount(pageCount){
		this.pageCount = pageCount;
	}


	/**
	 * Executes this operation synchronously using the supplied context and returns a new list of FileRef instances for the resultant PDF files.
	 *
	 * The resultant files may be stored in the system temporary directory (per the os.tempdir(), symlinks are resolved
	 * to the actual path).
	 * See {@link FileRef} for how temporary resources are cleaned up.
	 *
	 * @param {!ExecutionContext} context - The context in which the operation will be executed.
	 * @returns {Promise<FileRef[]>} A promise which resolves to the operation result.
	 * @throws {ServiceApiError} if an API call results in an error response.
	 * @throws {ServiceUsageError} if service usage limits have been reached or credentials quota has been exhausted.
	 */

	execute(context) {
		try {
			this.validate(context);
		} catch (err) {
			return Promise.reject(err);
		}
		logger.info('All validations successfully done. Beginning Split PDF operation execution');
		const targetFileName = getRandomFileNameWithExtension(allowedConfiguration.targetFileExtension),
			operationMessage = new OperationMessage(this.sourceFileRef, targetFileName, DefaultConfig.operationName.splitPDF),
			splitPDFService = new SplitPDFService();

		operationMessage.setOptions({
			pageRanges: this.pageRanges,
			pageCount: this.pageCount,
			fileCount: this.fileCount
		});
		return splitPDFService.perform(context, operationMessage)
			.then(res => Promise.resolve(res))
			.catch(err => Promise.reject(err));
	}

	validate(context) {
		validateClientContext(context);
		validateFileRef(this.sourceFileRef);
		validateAllowedMediaType(allowedConfiguration.getSupportedMediaTypes(), this.sourceFileRef);
		validateSplitOptions(this.pageRanges,this.pageCount,this.fileCount);
	}

}

Object.freeze(SplitPDFOperation);
module.exports = SplitPDFOperation;
