// @adobe/pdfservices-node-sdk/src/internal/auth/service-account-credentials-with-uri.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const ServiceAccountCredentials = require('../../auth/service-account-credentials'),
	DefaultConfig = require('../../internal/config/dc-services-default-config');


class ServiceAccountCredentialsWithUri extends ServiceAccountCredentials{

	constructor(ims_base_uri, client_id, client_secret, private_key, claim, organization_id, account_id,opsCreateUri) {
		super(client_id,client_secret,private_key,organization_id,account_id)
		this.imsBaseUri = ims_base_uri || DefaultConfig.imsBaseUri;
		this.claim = claim || (this.imsBaseUri + DefaultConfig.claim);
		this.opsCreateUri = opsCreateUri;
		this.validate();    											//parent (ServiceAccountCredentials) class validate method is called
	}

	getClaim() {
		return this.claim;
	}

	getImsBaseUri() {
		return this.imsBaseUri;
	}

	getOpsCreateUri() {
		return this.opsCreateUri;
	}
}

module.exports = ServiceAccountCredentialsWithUri
